### ISSUE TITLE
JavaClient: NPE causing all PartitionReceivers Receive pumps to exit and app to stop receiving messages

### ISSUE THREAD

**Issue by hanoch** [Azure/azure-event-hubs#181]
We are working with Mark Benyovszky and Lauren Mills on the Panama Canal project.

We updated the ArcGIS GeoEvent Azure connectors to build with the latest release of the “azure-eventhubs” SDK, version 0.7.2:
```xml
    <dependency>
      <groupId>com.microsoft.azure</groupId>
      <artifactId>azure-eventhubs</artifactId>
      <version>0.7.2</version>
    </dependency>
```

With this release we are experiencing a null pointer exception from the Partition Receiver (usually when there is no feed).  We are using the capability to start receiving events from the current time, using the API with the Date-Time parameter (and not the offset parameters):
```java
PartitionReceiver receiver = ehClient.createReceiver(EventHubClient.DEFAULT_CONSUMER_GROUP_NAME, Integer.toString(partitionId), Instant.now()).get();
```
(see the ArcGIS GeoEvent connector code - https://github.com/Esri/azure-for-geoevent/blob/master/azure-iot-hub-transport/src/main/java/com/esri/geoevent/transport/azure/AzureEventHubInboundTransport.java#L217)

This NPE is causing the Partition Receivers to close and stop receiving any events from there on:
```log
servicebus.trace	Receive pump for partition 3 exiting after user exception java.lang.NullPointerException	
servicebus.trace	Receive pump for partition 2 exiting after user exception java.lang.NullPointerException	
servicebus.trace	Receive pump for partition 1 exiting after user exception java.lang.NullPointerException	
servicebus.trace	Receive pump for partition 0 exiting after user exception java.lang.NullPointerException	
```

We are also noticing we are no longer able to receive new messages with this release of the SDK, using instant.now() API when creating the receivers (see above code snippet).

Setup:

azure-event-hubs version 0.7.2
Hub with four partitions
One EventHubClient with four PartitionReceivers
Attempting asynchronous reads of NEW messages, no writes

@SreeramGarlapati @velocipedist

**Comment by SreeramGarlapati** [comment]
@hanoch 
This exception/trace: ```
servicebus.trace    Receive pump for partition 3 exiting after user exception java.lang.NullPointerException    
```   is thrown when onReceive handler that is implemented by User code throws. Please refer to https://github.com/Azure/azure-event-hubs/blob/master/java/azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/PartitionReceiver.java#L393 

This is most likely a NPE in the EventHandler.onReceive. As a simple experiment - add a try-catch and log the exception stacktrace.

**Comment by hanoch** [comment]
my onEvents() handler method was not event invoked. Could it be that the TimeoutException logic is flowed and falls into one of the other exception handlers?

**Comment by hanoch** [comment]
@SreeramGarlapati 
it looks like the TimeoutException is raised in the PartitionReceiver class, which causes the NPE message.  I will send out a sample repro app.

**Comment by hanoch** [comment]
attached is a simple repo app:
[azure-for-geoevent-test.zip](https://github.com/Azure/azure-event-hubs/files/353011/azure-for-geoevent-test.zip)

**Comment by SreeramGarlapati** [comment]
@hanoch 
The sample you provided refers to <a href="http://mvnrepository.com/artifact/com.microsoft.azure.iothub-java-client/iothub-java-service-client/1.0.0-preview.9">iothub-java-service-client 1.0.0-preview.9</a> - which depends on a very old version of proton-j (0.10.0). azure-eventhubs 0.7.2 requires 0.13.0 to be loaded. When you have conflicting dependencies - the version loaded in non-deterministic in java.  This is the reason why you are running into NPE.

Remove the conflicting dependency and clear your local maven cache. This will be gone.

I ran the sample you provided by removing the iotclient dependency for >15 mins and cannot reproduce any problem.

These problems will comeout very vivid if you can print the StackTrace - the line numbers in the stack trace can help understand that a wrong version of jar was loaded.

**Comment by hanoch** [comment]
@SreeramGarlapati 
I removed the 'iothub-java-service-client' dependency, cleared the local repo cache and made sure the test app is build-with and loads 'prton-j' version 0.13.0.
My PartitionReceiverHandler.onReceive() method is still invoked with a null Iterable<EventData> value, after the PartitionReciver is getting a 'TimeoutException' exception:

https://github.com/Azure/azure-event-hubs/blob/master/java/azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/PartitionReceiver.java#L366

https://github.com/Azure/azure-event-hubs/blob/master/java/azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/PartitionReceiver.java#L402


As you requested, I added all the printStackTrace() statements in the repo app, and i am not seeing any client app code exception:
[azure-for-geoevent-test.zip](https://github.com/Azure/azure-event-hubs/files/357803/azure-for-geoevent-test.zip)

**Comment by hanoch** [comment]
@SreeramGarlapati 
The good news is that now the app is able to receive new messages (since 'NOW').
Please let us know which version of the 'iothub-java-service-client' SDK we can use in the same app, since we have the need to receive messages from an event hub, but also send events to event hubs and to Devices as well.

**Comment by SreeramGarlapati** [comment]
Hi @hanoch - am glad that you could get receive to work now. Did removing the iothub sdk dependency alone do the trick? 

Let me hook up with Iothub team for the correct version.

**Comment by hanoch** [comment]
I believe so.  I also added a check in my onReceive() method to return if invoked with a null variable.  No big deal, but I think you can clean this up as well in the PartitionReceiver code.  There is no need to invoke onReceive() with null on TimeOutException.

**Comment by hanoch** [comment]
adding a dependency on the latest released [iothub-java-service-client:1.0.7](http://mvnrepository.com/artifact/com.microsoft.azure.iothub-java-client/iothub-java-service-client/1.0.7) adds a dependency on [proton-j-azure-iot:0.12.2](http://mvnrepository.com/artifact/com.microsoft.azure.iot/proton-j-azure-iot/0.12.2)

**Comment by SreeramGarlapati** [comment]
Thanks for Quick respose @Hanoch.

1) Regarding the behavior of onReceive() invoked with null - this is a feature exposed by EventProcessorHost (Options -> invokeProcessorAfterReceiveTimeout).

2) I started a thread with IOTHub folks regarding the proton-j version question. As the proton-j-azure-iot is a private fork I couldn't find the answer myself. Lets track the issue there.

Am closing this as the current issue is resolved. Feel free to re-open if you feel the need.
